pkg/controller/direct/bigquerydataset/dataset_controller.go

```diff
@@ -0,0 +1,342 @@
+	log := klog.FromContext(ctx).WithName(ctrlName)
+	log.V(2).Info("getting BigQueryDataset", "name", a.id.External)
+
+	parent, datasetId, err := krm.ParseBigQueryDatasetExternal(a.id.External)
+	if err != nil {
+		return false, fmt.Errorf("failed to parse bigquery dataset full name, %w", err)
+	}
+	req := &bigquerypb.GetDatasetRequest{
+		DatasetId: &datasetId,
+		ProjectId: &parent.ProjectID,
+	}
+	datasetGetCall := a.gcpService.Datasets.Get(*req.ProjectId, *req.DatasetId)
+	datasetpb, err := datasetGetCall.Do()
+	if err != nil {
+		if direct.IsNotFound(err) {
+			return false, nil
+		}
+		return false, fmt.Errorf("getting BigQueryDataset %q: %w", a.id.External, err)
+	}
+	if err := convertAPIToProto(datasetpb, &a.actual); err != nil {
+		return false, nil
+	}
+	return true, nil
+}
+
+func (a *Adapter) Create(ctx context.Context, createOp *directbase.CreateOperation) error {
+
+	log := klog.FromContext(ctx).WithName(ctrlName)
+	log.V(2).Info("creating Dataset", "name", a.id.External)
+	mapCtx := &direct.MapContext{}
+
+	desiredDataset := BigQueryDataset_CovertKRMToAPI(mapCtx, a.desired)
+	desiredDataset.Labels = make(map[string]string)
+	for k, v := range a.desired.GetObjectMeta().GetLabels() {
+		desiredDataset.Labels[k] = v
+	}
+	desiredDataset.Labels["managed-by-cnrm"] = "true"
+	parent, _, err := krm.ParseBigQueryDatasetExternal(a.id.External)
+	if err != nil {
+		return fmt.Errorf("failed to parse bigquery dataset full name, %w", err)
+	}
+	// Resolve KMS key reference
+	if a.desired.Spec.DefaultEncryptionConfiguration != nil {
+		kmsRef, err := refs.ResolveKMSCryptoKeyRef(ctx, a.reader, a.desired, a.desired.Spec.DefaultEncryptionConfiguration.KmsKeyRef)
+		if err != nil {
+			return err
+		}
+		desiredDataset.DefaultEncryptionConfiguration.KmsKeyName = kmsRef.External
+	}
+	insertDatasetCall := a.gcpService.Datasets.Insert(parent.ProjectID, desiredDataset)
+	inserted, err := insertDatasetCall.Do()
+	if err != nil {
+		return fmt.Errorf("inserting Dataset %s: %w", a.id.External, err)
+	}
+	log.V(2).Info("successfully inserted Dataset", "name", a.id.External)
+
+	status := &krm.BigQueryDatasetStatus{}
+	status = BigQueryDatasetStatusObservedState_FromProto(mapCtx, inserted)
+	if mapCtx.Err() != nil {
+		return mapCtx.Err()
+	}
+	return createOp.UpdateStatus(ctx, status, nil)
+}
+
+func (a *Adapter) Update(ctx context.Context, updateOp *directbase.UpdateOperation) error {
+	u := updateOp.GetUnstructured()
+
+	log := klog.FromContext(ctx).WithName(ctrlName)
+	log.V(2).Info("updating Dataset", "name", a.id.External)
+	mapCtx := &direct.MapContext{}
+
+	// Convert KRM object to proto message
+	desiredKRM := a.desired.DeepCopy()
+	desired := BigQueryDatasetSpec_ToProto(mapCtx, &desiredKRM.Spec, desiredKRM.Name)
+	if mapCtx.Err() != nil {
+		return mapCtx.Err()
+	}
+	actual := a.actual
+	resource := proto.Clone(a.actual).(*bigquerypb.Dataset) // this is the proto resource we are passing to GCP API update call.
```

Comment:
Do we need a `bigquerypb` resource? Conceptually the controller only need to convert between KRM and API, like the implementation of the Create() function.

Reply:
Hmm technically yes, I can remove the `bigquerypb` resource conversion.